### PR TITLE
trie, eth/protocols/snap: less tiny db-writes during trie-healing

### PR DESCRIPTION
**Pull request by holiman**
This PR tries to prevent making tiny writes, by only performing the batch-write if the accumulated data is large enough to be meaningful.

The current code, at worst case, would make a tiny batch-write on each trienode response.

### CODE REVIEW

File: trie/sync.go [hunk 1]
```diff
@@ -337,6 +338,10 @@ func (s *Sync) Commit(dbw ethdb.Batch) error {
 	return nil
 }
 
+func (s *Sync) MemSize() uint64 {
```
Comment (holiman) [inline trie/sync.go:341]: docstring

File: trie/sync.go [hunk 2]
```diff
@@ -110,6 +110,7 @@ type syncMemBatch struct {
 	nodes  map[string][]byte      // In-memory membatch of recently completed nodes
 	hashes map[string]common.Hash // Hashes of recently completed nodes
 	codes  map[common.Hash][]byte // In-memory membatch of recently completed codes
+	size   uint64
```
Comment (holiman) [inline trie/sync.go:113]: docstring

File: eth/protocols/snap/sync.go [hunk 3]
```diff
@@ -2121,6 +2129,7 @@ func (s *Syncer) processTrienodeHealResponse(res *trienodeHealResponse) {
 		log.Crit("Failed to persist healing data", "err", err)
 	}
 	log.Debug("Persisted set of healing data", "type", "trienodes", "bytes", common.StorageSize(batch.ValueSize()))
+
```
Comment (holiman) [inline eth/protocols/snap/sync.go:2132]: rm

File: trie/sync.go [hunk 4]
```diff
@@ -458,7 +463,7 @@ func (s *Sync) commitNodeRequest(req *nodeRequest) error {
 	// Write the node content to the membatch
 	s.membatch.nodes[string(req.path)] = req.data
 	s.membatch.hashes[string(req.path)] = req.hash
-
```
Comment (MariusVanDerWijden) [inline trie/sync.go:461]: Could you doc this too? Very hard to follow and easy to miss if we ever need to update the way a node is stored

Comment (rjl493456442) [reply trie/sync.go:461]: Yeah I agree. I think in hash-based, we don't need to count the req.path since it's not a part of storage data. Maybe just `common.HashLength + len(req.data)`?

And we need to introduce the state scheme abstraction here. Like the PR https://github.com/ethereum/go-ethereum/pull/25532/files I can add one more API called `NodeSize` to return the approximate node size according to the scheme we use.

File: eth/protocols/snap/sync.go [hunk 5]
```diff
@@ -620,6 +620,7 @@ func (s *Syncer) Sync(root common.Hash, cancel chan struct{}) error {
 		trienodeHealResps    = make(chan *trienodeHealResponse)
 		bytecodeHealResps    = make(chan *bytecodeHealResponse)
 	)
+	defer s.commit(true)
```
Comment (rjl493456442) [inline eth/protocols/snap/sync.go:623]: We can put it after Line 586, https://github.com/ethereum/go-ethereum/pull/25681/files#diff-d26ea400bc58a59ffc652f18f3e402fe7e3279c1569c8374c216b5d702f2fe7cR586

It's cleaner for all persistence defer logics.

**Review by rjl493456442 — APPROVED** [review]

**Comment by MariusVanDerWijden** [comment]
This sounds good to me, my only question would be if we have the same ordering requirement that we have with the trie expansion thingy which created the bad blocks in the 1.10.23 release

**Comment by holiman** [comment]
> if we have the same ordering requirement that we have with the trie expansion thingy

We do, and we don't. At this point, the ordering has already happened, and what's being written is nodes from the bottom-up. This PR simply batches it a bit. This will not change the order of the writes, just defer them a bit. In the event of a crash, and the writes get lost, there's no real difference from before. 

Aside from writing in a more meaningful size, this also means that the `membatch` becomes a bit more useful as a lookup.

**Review by rjl493456442 — APPROVED** [review]
I think it should be good to go!
